timelock-corruption-detection/src/main/java/com/palantir/history/PaxosLogHistoryProvider.java

```diff
@@ -61,16 +80,131 @@ private Long getOrInsertVerificationState(NamespaceAndUseCase namespaceAndUseCas
                 namespaceAndUseCase.namespace(), namespaceAndUseCase.useCase());
     }
 
-    public List<CompletePaxosHistoryForNamespaceAndUseCase> getHistory() {
+    /**
+     * Gets history from all the nodes for all unique namespace, useCase tuples since last verified sequence numbers
+     *
+     * @throws  RuntimeException
+     *          if fails to fetch history from all remote servers
+     */
+    public List<CompletePaxosHistoryForNamespaceAndUseCase> getHistory() throws RuntimeException {
+
+        /* used to load history since last verified sequence numbers for all unique namespace and use case tuples */
         Map<NamespaceAndUseCase, Long> lastVerifiedSequences = KeyedStream
                 .of(getNamespaceAndUseCaseTuples().stream())
                 .map(namespaceAndUseCase -> verificationProgressStateCache.computeIfAbsent(namespaceAndUseCase,
                         this::getOrInsertVerificationState))
                 .collectToMap();
+
+        /**
+         * The history queries are built from the lastVerifiedSequences map above,
+         * required for conjure endpoints to load remote history
+         */
+        List<HistoryQuery> historyQueries = KeyedStream.stream(lastVerifiedSequences)
+                .mapEntries(this::buildHistoryQuery)
+                .values()
+                .collect(Collectors.toList());
+
         PaxosHistoryOnSingleNode localPaxosHistory = localHistoryLoader.getLocalPaxosHistory(
                 lastVerifiedSequences);
 
-        // TBD get history from remotes, merge, return
-        return ImmutableList.of();
+        /**
+         * List of logs from all remotes
+         */
+        List<List<LogsForNamespaceAndUseCase>> rawHistoryFromAllRemotes = remoteHistoryProviders.stream().map(
+                remote -> fetchHistoryFromRemote(historyQueries, remote)).collect(Collectors.toList());
+
+        /**
+         * List of history from all remotes. Each history is map of namespaceAndUseCase to Paxos logs. The Paxos logs
+         * are mapped against sequence numbers.
+         */
+        List<Map<NamespaceAndUseCase, Map<Long, LearnedAndAcceptedValue>>> historyLogsFromRemotes
```

Comment:
I think we should have some wrapper types here, the primitives get tricky to read after a while.